### PR TITLE
Updates Tz Singletons and Caches with `WeakValueDictionary`s

### LINKED ISSUE #635: Switch timezone singletons and caches with weak references
Right now `tzoffset`, `tzstr` and `tz.gettz` each retain strong references to every unique time zone ever constructed by that method. I think in the end this will be a net savings in memory in *most* use cases since it prevents duplication of time zones, but it could also cause a huge proliferation of memory in applications that make short-lived references to a large number of time zones (since those time zones, when they go out of scope, would normally get garbage collected).

In Python 3 there is the [weakref](https://docs.python.org/3/library/weakref.html) module, which allows the creation of weak references, such that an object will be garbage collected if only weak references to it exist.

Rather than try and use `backports.weakref` in Python 2, I think we can just write a "pass through" compatibility shim that has the same external form as `weakref` but simply holds a strong reference to the object in question.

### PR DESCRIPTION
**Pull request by cs-cordero**
Closes #635.

### CODE REVIEW

File: dateutil/utils.py [hunk 1]
```diff
@@ -4,6 +4,10 @@
 from datetime import datetime, time
 
 
+class FakeWeakRef:
```
Comment (pganssle) [inline dateutil/utils.py]: `.utils` is a public module and very much *not* the right place for this. It should probably be in `tz._common.py`

File: dateutil/tz/_factories.py [hunk 2]
```diff
@@ -1,5 +1,10 @@
 from datetime import timedelta
 
+try:
+    import weakref
```
Comment (pganssle) [inline dateutil/tz/_factories.py]: I would change this to:

```python
try:
    from weakref import WeakRefDictionary
except ImportError:
    WeakRefDictionary = dict
```

File: dateutil/test/test_tz.py [hunk 3]
```diff
@@ -14,10 +14,12 @@
 import sys
 import base64
 import copy
+import gc
 
 from functools import partial
 
 IS_WIN = sys.platform.startswith('win')
+NOT_PY3 = not sys.version_info[0] == 3
```
Comment (pganssle) [inline dateutil/test/test_tz.py]: Normally you would not have to define this. There's already a `six.PY3` and `six.PY2` that will give you that.

That said, I may have been wrong about weak references being a Python 3 thing. [Seems like they are available in Python 2.7](https://docs.python.org/2.7/library/weakref.html), so we should drop this skip test.

File: dateutil/tz/_factories.py (hunk 2 above)
Comment (pganssle) [reply dateutil/tz/_factories.py]: Actually it turns out that Python 2 has weak references, so we can just drop the whole compatibility shim. My bad.

**Comment by pganssle** [comment]
Hard to tell if this is just holding a strong reference somewhere we don't realize or if the assumptions about one of the inputs is wrong. Seems to only happen on Windows on old versions of Python, though.

**Comment by cs-cordero** [comment]
Sorry for the flurry of seemingly random commits.  I can't seem to replicate AppVeyor's failure on my local machine even after matching the Python version, so I'm trying to troubleshoot with small changes.

**Comment by cs-cordero** [comment]
Just reverted all the troubleshooting commits, but I'm kind of at a loss here for how this is failing on the AppVeyor CI.

The stdout from the 46bdab7 commit is really what's really tripping me up:
https://ci.appveyor.com/project/dateutil/dateutil/build/2.3.1234/job/4ru7kdg7wipal15w

In this commit, I print the keys inside the `WeakValueDictionary` before and after deleting the strong `ATK1` and `ATK2` references, expecting the WeakValueDictionary to clear itself, but it does not.  I also used `pytest.raises(NameError)` to assert that `ATK1` and `ATK2` are indeed deleted, and they are.  There shouldn't be any remaining strong references to the value in the dictionary, but somehow it is persisting.  Running the equivalent code locally off of CI yields the expected behavior.

I'll keep thinking about how to get around this.  One idea I had was to have some kind of `_cache_hit` attribute on the factory that keeps a count of every time it is returning a value from its `WeakValueDictionary`.  We can assert that this counter does/does not change depending on the test.

**Comment by pganssle** [comment]
I'm guessing that this is a real issue with Windows and not a problem with the tests. I'll grab this branch and try debugging it on a Windows VM.

**Comment by pganssle** [comment]
Cleaned up the commit history. Will merge when tests pass.

Thanks so much @cs-cordero!

**Comment by cs-cordero** [comment]
Happy to help. Thanks for taking it home with that clutch Windows `ZoneInfoFile` issue.

**Comment by pganssle** [comment]
Hm.. The Python 3.5 test failed but succeeded on reboot. This "id won't get reused" gambit may be less stable than I hoped.

I just had a new idea for a more robust test, though. Create a new weak reference to the object before deleting it and make sure its referrent is gone after the gc step.

**Comment by pganssle** [comment]
@cs-cordero One last thing, can you add yourself to `AUTHORS.md` - names are in alphabetical order. You are welcome to put your real e-mail and encouraged to put github name. Make sure to put a **D** after your name line, which indicates your contributions are under the dual license.

**Comment by cs-cordero** [comment]
Will do re: `authors.md` once I get access to a computer and little later in the day. Thanks!

**Comment by pganssle** [comment]
@cs-cordero No this test assumes that the `weakref` module works as expected and tests that we didn't screw up and accidentally start holding a strong reference to the `tzinfo` objects in the cache.

If you do:

```python
git checkout master
git checkout c610fc509674a -- dateutil/test/test_tz.py
pytest -k test_tz.py
```

You'll see that these three tests fail, which means the problem - that the cache itself holds a strong reference to the cached time zones - is solved by this patch.

**Comment by cs-cordero** [comment]
Ah yes that makes perfect sense! Thank you for explaining.

**Comment by pganssle** [comment]
@cs-cordero You'll probably need to do a force pull when you edit this branch again, because I've rebased your branch and rewritten the history. If you don't know how to do that it's `git pull --force`

**Comment by cs-cordero** [comment]
🎉

**Pull request merged**